src/source.cpp

```diff
@@ -322,6 +352,135 @@ Particle::Bank sample_external_source()
+  return sampler;
+}
+
+Particle::Bank sample_pyne_source(pyne::Sampler* sampler)
+{
+  // Set the random number generator to the source stream.
+  prn_set_stream(STREAM_SOURCE);
+
+  // Determine total source strength
+  double total_strength = 1.0;
+
+  // Repeat sampling source location until a good site has been found
+  bool found = false;
+  int n_reject = 0;
+  int max_void_rejection = 100;
+  std::vector<double> rands;
+  pyne::SourceParticle src;
+  Particle::Bank site;
+  // step 1: There are 6 random numbers used to sample the
+  //         mesh element index, energy, x, y, z, weight
+  //         1.1 sample mesh element index using rands[0] and rands[1]
+  //         1.2 sample particle location x, y, z using rands[2, 3, 4]
+  //         1.3 sample energy
+  // step 2: check the location of the particle (in the geom? void?)
+  // step 3: If this partilce failed in the check, repeat from 1.2 to try
+  //         another location x, y, z.
+  //         If failed max_void_rejection times, repeat from 1.1 to try another
+  //         mesh element & energy.
+  // step 4: If particle passes the check. Done and convert it to openmc format.
+  while (!found) {
+    // Generate rands used in particle_birth
+    // For each mesh element, try at most max_void_rejection times
+    if (!found and (n_reject == 0 or n_reject == max_void_rejection)) {
+      // rands[0] and rands[1] are used for alias table, determin the mesh index
+      // rands[6] for energy
+      // rands[2], rands[3] and rands[4] for x, y, and z
+      // initialize the rands at the beginning for each mesh element
+      rands.clear();
+      for (int i=0; i<6; i++){
+          rands.push_back(prn());
+      }
+      // reset n_reject and sample particle in new mesh element
+      n_reject = 0;
+    }
+    if (!found and n_reject != 0) {
+      // update the rands for x, y, z
+      rands[2] = prn();
+      rands[3] = prn();
+      rands[4] = prn();
+    }
+    // Sample particle
+    src = sampler->particle_birth(rands);
+    site = convert_pyne_source_particle(src);
+    // Check for rejection
+    found = check_pyne_source_particle(site);
+    if (!found) {
+      ++n_reject;
+    }
+  }
+
+  // If running in MG, convert site % E to group
+  if (!settings::run_CE) {
+    site.E = lower_bound_index(data::rev_energy_bins.begin(),
+      data::rev_energy_bins.end(), site.E);
+    site.E = data::num_energy_groups - site.E;
+  }
+
+  // Set the random number generator back to the tracking stream.
+  prn_set_stream(STREAM_TRACKING);
+
+  return site;
+}
+
+Particle::Bank convert_pyne_source_particle(pyne::SourceParticle pyne_src)
+{
+  Particle::Bank site;
+  site.r = Position(pyne_src.get_x(), pyne_src.get_y(), pyne_src.get_z());
+
+  // Sample angle
+  Isotropic angle;
+  site.u = angle_->sample();
+  site.E = pyne_src.get_e();
+  site.wgt = pyne_src.get_w();
+  site.delayed_group = 0;
+  // Check for particle type
+  if (!settings::photon_transport) {
```

Comment:
I'm not sure this is the right way to handle this situation. Unless I'm missing something, the particle generated by the source should depend on the source specification, not on whether or not a type of particle is enabled in the transport. If a user is asking for a PyNE photon source and disables photon transport in OpenMC we should report an error rather than generate neutrons here.

Reply:
This is partially why I think a dedicated `SourceDistribution` class for the PyNE source might be a good idea.

Reply:
Agreed, source particle type and the transport type are independent. I will add a parameter to define it explicitly.